core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -142,6 +147,52 @@ public static PositionDeleteIndex toPositionIndex(CloseableIterable<Long> posDel
     }
   }
 
+  public static <T extends StructLike> Map<String, PositionDeleteIndex> toPositionIndexMap(
+      CharSequenceSet dataLocations, List<CloseableIterable<T>> deleteFiles) {
+    DataFileFilter<T> locationFilter = new DataFileFilter<>(dataLocations);
+    List<CloseableIterable<Pair<String, Long>>> positions =
+        Lists.transform(
+            deleteFiles,
+            deletes ->
+                CloseableIterable.transform(
+                    locationFilter.filter(deletes),
+                    row ->
+                        Pair.of(
+                            (String) FILENAME_ACCESSOR.get(row),
+                            (Long) POSITION_ACCESSOR.get(row))));
+
+    try (CloseableIterable<Pair<String, Long>> deletes = CloseableIterable.concat(positions)) {
+      Map<String, PositionDeleteIndex> positionDeleteIndex = Maps.newHashMap();
+      deletes.forEach(
+          entry ->
+              positionDeleteIndex
+                  .computeIfAbsent(entry.first(), f -> new BitmapPositionDeleteIndex())
```

Comment:
Instead of using `computeIfAbsent` on every row, this should pre-populate the map using `dataLocations`, since those are all known.